src/cluster.c

```diff
@@ -5808,11 +5920,8 @@ NULL
         }
     } else if (!strcasecmp(c->argv[1]->ptr,"nodes") && c->argc == 2) {
         /* CLUSTER NODES */
-        /* Report plaintext ports, only if cluster is TLS but client is known to
-         * be non-TLS). */
-        int use_pport = (server.tls_cluster &&
-                        c->conn && (c->conn->type != connectionTypeTls()));
-        sds nodes = clusterGenNodesDescription(c, 0, use_pport);
+        /* Report TLS ports to TLS client, and report non-TLS port to non-TLS client. */
+        sds nodes = clusterGenNodesDescription(c, 0, connIsTLS(c->conn));
```

Comment:
@madolson @CharlesChen888 This change broke the reply of `cluster slots` when it is invoke from Lua or from a module using `RM_Call`. It is always return the none tls port:
```
127.0.0.1:6379> eval "return redis.call('cluster', 'slots')" 0
1) 1) (integer) 0
   2) (integer) 8192
   3) 1) "127.0.0.1"
      2) (integer) 0
      3) "07b88e24f8ad4de988d03d1679646756a132e2ab"
      4) (empty array)
2) 1) (integer) 8193
   2) (integer) 16383
   3) 1) "127.0.0.1"
      2) (integer) 0
      3) "58fcfa7ec6b78d70f20d5c25dfd3eda56debb811"
      4) (empty array)
```

 For modules it is possible to use `RM_GetClusterNodeInfo` (though an existing modules that uses `RM_Call` with `cluster slots` command will break). For Lua I do not see a workaround.
 
 Do you see it as a breaking change? And if so, should we fix it for both modules and Lua or maybe only Lua?

@yossigo @oranagra FYI

Reply:
> i suppose there's no good reason for Lua to call this command

I have a feeling it's dangerous to assume things like that. There is no limitation to which commands can be used from Lua, for administration and who knows what, so I think we should at least try to keep backward compatibility, if it's not too complex anyway.

So, the client has no connection (`c->conn == NULL`). Before this PR, the default was depending on `server.tls_cluster` but now we default to non-TLS (we return TLS port if and only if `conn && conn->type == connectionTypeTls()`).

I suggest we fix it in one of these ways:

1. fallback to `server.tls_cluster` when the client has no connection, i.e. change `connIsTLS` to return `conn ? conn->type == connectionTypeTls() : server.tls_cluster`, or
2. look at `c->flags & (CLIENT_SCRIPT|CLIENT_MODULE)` and use `server.current_client` to determine TLS or not.

Reply:
Perhaps the fake client in Lua and RM_Call can somehow inherit some connection properties from the real client?

Reply:
> fallback to server.tls_cluster when the client has no connection

I don't think this is a proper way. In this case, if a client connects through TLS but tls-cluster is set to 0, the client is supposed to get TLS ports but server will return TCP port.

Reply:
Also notice that on modules there might not be a real client, module might operate on background thread or from cron or `RM_CreateTimer` maybe.

Reply:
i suppose we should always use `server.current_client` (if exists) instead of `c`. and if that one has no `conn`, then fallback to `server.tls_cluster`.
this approach probably doesn't have any drawbacks compared to what we have now, and is more backwards compatible with what we had before.

Reply:
> That's right, but it was the exact behaviour before this PR, so if the main priority is to avoid breaking changes, this is an option.

I don't really disagree, but I also don't think this was defined behavior and have no real concerns about changing it. Cluster slots is intended for clients to use and has behavior that relates to how the client connected. I agree about using the server.current_client when possible, but think we should return null for the cron case as it's really undefined behavior. If you want defined behavior, use cluster shards.

Reply:
@MeirShpilraien please make a PR and we'll continue to discuss it there.